fileServer.js

```diff
@@ -0,0 +1,67 @@
+ 'use strict';
+
+ var http = require('http');
+ var fs = require('fs');
+ var path = require('path');
+ var url = require('url');
+ var rootDir = __dirname + '/public';
+
+ const mimeType = {
+   'html': 'text/html',
+   'jpeg': 'image/jpeg',
+   'jpg': 'image/jpeg',
+   'png': 'image/png',
+   'js': 'text/javascript',
+   'css': 'text/css'
+ };
+
+ function send404Request(res) {
+   fs.readFile('public/404.html', 'UTF-8', function(err, html) {
+     res.setHeader('Content-type', mimeType);
+     res.end(html);
+   });
+ }
+
+ function onRequest(req, res) {
+   console.log('request starting...');
+
+   var filename;
+   var pathname = url.parse(req.url).pathname;
+   filename = path.join(rootDir, pathname);
+   console.log('Filename: ' + filename);
+
+   if (req.method !== 'GET') {
+     res.writeHead(405, {
+       "Content-type": "text/plain"
+     });
+     res.write(`HTTP method ${req.method} not yet supported`);
+     res.end();
+     return
+   }
+
+   if (req.url === '/') {
+     res.writeHead(200, {
+       'Content-type': mimeType
+     });
+     fs.createReadStream('public/index.html').pipe(res);
+   }
+
+   fs.exists(filename, function(exists) {
```

Comment:
Only checking for file existence is not a guarantee the file will be readable, files have permissions and it could be the file is there but not readable.

Also, seeing from your code, you check first if it exists, **then** you attempt to read it - this could result in a race condition where between the time the code checks if the code exists and the readFile function is called, the file has been deleted or is otherwise inaccessible.

A preferable pattern from a systems programming perspective is this (pseudo-code, not javascript)

```
contents =  read file using read function, passing full path and name of tile
if error occrred while reading file
   log error for operations / developer, include full error message from system
   return 404 to user
else
   set content type based on file extension
   pipe contents to HTTP response object
   end response
end